# Benchmark Summary Statistics
---

Description:

    This notebook performs the summary statisticks results for all benchmark functions (one at a time).
    The optimization process is repeated N=50 times (with default settings) and then we collect the results.
    
---

### Import python libraries and set up the project directory

In [6]:
import os, sys
import numpy as np

PROJECT_DIR = os.path.abspath('..')
sys.path.append(PROJECT_DIR)

### Import the custom PSO code

In [7]:
from star_pso.population.swarm import Swarm
from star_pso.population.particle import Particle
from star_pso.engines.standard_pso import StandardPSO
from star_pso.utils.auxiliary import cost_function

### Load all the Benchmark problems

In [8]:
# Dimensions: 1D
from star_pso.benchmarks.equal_maxima import EqualMaxima
from star_pso.benchmarks.five_uneven_peak_trap import FiveUnevenPeakTrap
from star_pso.benchmarks.uneven_decreasing_maxima import UnevenDecreasingMaxima

# Dimensions: 2D
from star_pso.benchmarks.himmelblau import Himmelblau
from star_pso.benchmarks.six_hump_camel_back import SixHumpCamelBack

# Dimensions: N >= 2D
from star_pso.benchmarks.shubert import Shubert
from star_pso.benchmarks.vincent import Vincent
from star_pso.benchmarks.rastrigin import Rastrigin

In [9]:
# Select one test problem.
benchmark = Shubert()

### Define the cost function

Note: All the optimization problems are solved as maximizations.

In [10]:
@cost_function(minimize=False)
def func_benchmark(x_array: np.ndarray, **kwargs) -> float:
    # Return the solution.
    return benchmark.func(x_array).item()
# _end_def_

### Setup the experiment

In [11]:
# Number of repeatitions.
n_repeats = 50

# Swarm population size.
POP_SIZE = 500

# Maximum function evaluations.
MAX_F_EVALS = 150_000

# Store the results.
results = n_repeats * [0]

# Display Info.
print(f"Testing function: {benchmark.name}\n")

# Go time ...
for i in range(n_repeats):

    # Display current iteration.
    print(f"Running optimization: {i+1} out of {n_repeats} ... ", flush=True)

    # Get an initial (new) sample.
    x_sample = benchmark.sample_random_positions(n_pos=POP_SIZE)

    # Initial swarm population.
    swarm_t0 = Swarm([Particle(x) for x in x_sample])
    
    # Create the StandardPSO object that will carry on the optimization.
    test_PSO = StandardPSO(initial_swarm = swarm_t0, obj_func = func_benchmark,
                           x_min = benchmark.x_min, x_max = benchmark.x_max)
    # Run the optimization.
    test_PSO.run(max_it = 800, f_max_eval = MAX_F_EVALS, verbose = False,
                 options = {"w0": 0.75, "c1": 1.50, "c2": 1.50, "mode": "multimodal"})
    
    # Obtain the number of the global optima that were found for epsilon = 1.0e-4.
    gopt_found, gopt_total = benchmark.search_for_optima(test_PSO.swarm.best_n(POP_SIZE))

    # Update the results.
    results[i] = gopt_found

    # Blank line to declatter the screen.
    print(" ")
# _end_for_

print("Done.")

Testing function: Shubert_2D

Running optimization: 1 out of 50 ... 


11/28/2025 16:19:21 INFO: Initial f_optimal = 151.9789
11/28/2025 16:19:38 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 17.318 seconds.
 
Running optimization: 2 out of 50 ... 


11/28/2025 16:19:38 INFO: Initial f_optimal = 183.3318
11/28/2025 16:19:49 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.509 seconds.
 
Running optimization: 3 out of 50 ... 


11/28/2025 16:19:49 INFO: Initial f_optimal = 178.4418
11/28/2025 16:20:01 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.577 seconds.
 
Running optimization: 4 out of 50 ... 


11/28/2025 16:20:01 INFO: Initial f_optimal = 147.7994
11/28/2025 16:20:12 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.506 seconds.
 
Running optimization: 5 out of 50 ... 


11/28/2025 16:20:13 INFO: Initial f_optimal = 137.8471
11/28/2025 16:20:24 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.569 seconds.
 
Running optimization: 6 out of 50 ... 


11/28/2025 16:20:24 INFO: Initial f_optimal = 180.2078
11/28/2025 16:20:35 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.366 seconds.
 
Running optimization: 7 out of 50 ... 


11/28/2025 16:20:36 INFO: Initial f_optimal = 177.7455
11/28/2025 16:20:47 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.304 seconds.
 
Running optimization: 8 out of 50 ... 


11/28/2025 16:20:47 INFO: Initial f_optimal = 185.8398
11/28/2025 16:20:59 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 12.101 seconds.
 
Running optimization: 9 out of 50 ... 


11/28/2025 16:20:59 INFO: Initial f_optimal = 183.6897
11/28/2025 16:21:10 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.543 seconds.
 
Running optimization: 10 out of 50 ... 


11/28/2025 16:21:11 INFO: Initial f_optimal = 133.7477
11/28/2025 16:21:22 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.498 seconds.
 
Running optimization: 11 out of 50 ... 


11/28/2025 16:21:22 INFO: Initial f_optimal = 172.3960
11/28/2025 16:21:34 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.523 seconds.
 
Running optimization: 12 out of 50 ... 


11/28/2025 16:21:34 INFO: Initial f_optimal = 79.0154
11/28/2025 16:21:45 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.418 seconds.
 
Running optimization: 13 out of 50 ... 


11/28/2025 16:21:45 INFO: Initial f_optimal = 163.1198
11/28/2025 16:21:56 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.465 seconds.
 
Running optimization: 14 out of 50 ... 


11/28/2025 16:21:57 INFO: Initial f_optimal = 175.1565
11/28/2025 16:22:08 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.589 seconds.
 
Running optimization: 15 out of 50 ... 


11/28/2025 16:22:08 INFO: Initial f_optimal = 128.3116
11/28/2025 16:22:20 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.628 seconds.
 
Running optimization: 16 out of 50 ... 


11/28/2025 16:22:20 INFO: Initial f_optimal = 106.4553
11/28/2025 16:22:31 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.653 seconds.
 
Running optimization: 17 out of 50 ... 


11/28/2025 16:22:31 INFO: Initial f_optimal = 165.1499
11/28/2025 16:22:43 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.634 seconds.
 
Running optimization: 18 out of 50 ... 


11/28/2025 16:22:43 INFO: Initial f_optimal = 163.8710
11/28/2025 16:22:55 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.681 seconds.
 
Running optimization: 19 out of 50 ... 


11/28/2025 16:22:55 INFO: Initial f_optimal = 178.0672
11/28/2025 16:23:06 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.545 seconds.
 
Running optimization: 20 out of 50 ... 


11/28/2025 16:23:06 INFO: Initial f_optimal = 171.4280
11/28/2025 16:23:18 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.761 seconds.
 
Running optimization: 21 out of 50 ... 


11/28/2025 16:23:18 INFO: Initial f_optimal = 180.6592
11/28/2025 16:23:30 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.737 seconds.
 
Running optimization: 22 out of 50 ... 


11/28/2025 16:23:30 INFO: Initial f_optimal = 157.5259
11/28/2025 16:23:42 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.680 seconds.
 
Running optimization: 23 out of 50 ... 


11/28/2025 16:23:42 INFO: Initial f_optimal = 159.8492
11/28/2025 16:23:53 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.713 seconds.
 
Running optimization: 24 out of 50 ... 


11/28/2025 16:23:53 INFO: Initial f_optimal = 166.1284
11/28/2025 16:24:05 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.678 seconds.
 
Running optimization: 25 out of 50 ... 


11/28/2025 16:24:05 INFO: Initial f_optimal = 176.2367
11/28/2025 16:24:16 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.349 seconds.
 
Running optimization: 26 out of 50 ... 


11/28/2025 16:24:16 INFO: Initial f_optimal = 177.8963
11/28/2025 16:24:28 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.307 seconds.
 
Running optimization: 27 out of 50 ... 


11/28/2025 16:24:28 INFO: Initial f_optimal = 98.8290
11/28/2025 16:24:39 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.314 seconds.
 
Running optimization: 28 out of 50 ... 


11/28/2025 16:24:39 INFO: Initial f_optimal = 158.9275
11/28/2025 16:24:50 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.408 seconds.
 
Running optimization: 29 out of 50 ... 


11/28/2025 16:24:51 INFO: Initial f_optimal = 103.5654
11/28/2025 16:25:02 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.339 seconds.
 
Running optimization: 30 out of 50 ... 


11/28/2025 16:25:02 INFO: Initial f_optimal = 144.6538
11/28/2025 16:25:13 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.301 seconds.
 
Running optimization: 31 out of 50 ... 


11/28/2025 16:25:13 INFO: Initial f_optimal = 186.6196
11/28/2025 16:25:25 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.322 seconds.
 
Running optimization: 32 out of 50 ... 


11/28/2025 16:25:25 INFO: Initial f_optimal = 182.1734
11/28/2025 16:25:36 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.359 seconds.
 
Running optimization: 33 out of 50 ... 


11/28/2025 16:25:36 INFO: Initial f_optimal = 137.2352
11/28/2025 16:25:47 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.286 seconds.
 
Running optimization: 34 out of 50 ... 


11/28/2025 16:25:47 INFO: Initial f_optimal = 182.5691
11/28/2025 16:25:59 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.320 seconds.
 
Running optimization: 35 out of 50 ... 


11/28/2025 16:25:59 INFO: Initial f_optimal = 184.1676
11/28/2025 16:26:10 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.323 seconds.
 
Running optimization: 36 out of 50 ... 


11/28/2025 16:26:10 INFO: Initial f_optimal = 163.4625
11/28/2025 16:26:21 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.248 seconds.
 
Running optimization: 37 out of 50 ... 


11/28/2025 16:26:21 INFO: Initial f_optimal = 185.3039
11/28/2025 16:26:33 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.386 seconds.
 
Running optimization: 38 out of 50 ... 


11/28/2025 16:26:33 INFO: Initial f_optimal = 121.4703
11/28/2025 16:26:44 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.396 seconds.
 
Running optimization: 39 out of 50 ... 


11/28/2025 16:26:44 INFO: Initial f_optimal = 139.0069
11/28/2025 16:26:55 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.245 seconds.
 
Running optimization: 40 out of 50 ... 


11/28/2025 16:26:55 INFO: Initial f_optimal = 177.2873
11/28/2025 16:27:07 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.283 seconds.
 
Running optimization: 41 out of 50 ... 


11/28/2025 16:27:07 INFO: Initial f_optimal = 129.2111
11/28/2025 16:27:18 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.363 seconds.
 
Running optimization: 42 out of 50 ... 


11/28/2025 16:27:18 INFO: Initial f_optimal = 166.5330
11/28/2025 16:27:29 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.476 seconds.
 
Running optimization: 43 out of 50 ... 


11/28/2025 16:27:29 INFO: Initial f_optimal = 153.0487
11/28/2025 16:27:41 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.357 seconds.
 
Running optimization: 44 out of 50 ... 


11/28/2025 16:27:41 INFO: Initial f_optimal = 177.8983
11/28/2025 16:27:52 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.291 seconds.
 
Running optimization: 45 out of 50 ... 


11/28/2025 16:27:52 INFO: Initial f_optimal = 133.2947
11/28/2025 16:28:04 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.351 seconds.
 
Running optimization: 46 out of 50 ... 


11/28/2025 16:28:04 INFO: Initial f_optimal = 183.6968
11/28/2025 16:28:15 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.302 seconds.
 
Running optimization: 47 out of 50 ... 


11/28/2025 16:28:15 INFO: Initial f_optimal = 146.2566
11/28/2025 16:28:26 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.393 seconds.
 
Running optimization: 48 out of 50 ... 


11/28/2025 16:28:26 INFO: Initial f_optimal = 125.0355
11/28/2025 16:28:38 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.347 seconds.
 
Running optimization: 49 out of 50 ... 


11/28/2025 16:28:38 INFO: Initial f_optimal = 179.5105
11/28/2025 16:28:49 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.321 seconds.
 
Running optimization: 50 out of 50 ... 


11/28/2025 16:28:49 INFO: Initial f_optimal = 150.8868
11/28/2025 16:29:00 WARNING: StandardPSO reached the maximum number of function evaluations at iteration 299


Final f_optimal = 186.7309
run: elapsed time = 11.391 seconds.
 
Done.


In [12]:
# Compute the "Peak Ratio".
peak_ratio = sum(results) / (gopt_total * n_repeats)

print(f"PR = {peak_ratio}")

PR = 0.9588888888888889


In [13]:
# Compute the "Success Rate".
success_rate = sum([r == gopt_total for r in results]) / n_repeats

print(f"SR = {success_rate}")

SR = 0.48


### Optimization Results

Accuracy level: $\epsilon = 1.0^{-4}$

| Benchmark Function | Dimensions | No. Global opt. | MaxFEs  | Population Size | PR | SR |
|:--|:--|:--|:--|:--|:--|:--|
| Equal Maxima | 1 | 5 | 50_000| 125 | 1.0 | 1.0 |
| Five Uneven Peak Trap | 1 | 2 | 50_000 | 125 | 1.0 | 1.0 |
| Uneven Decreasing Maxima | 1 | 1 | 50_000 | 85 | 1.0 | 1.0 |
| Himmelblau | 2 | 4 | 50_000 | 100 | 1.0 | 1.0 |
| Six Hump Camel Back | 2 | 2 | 50_000 | 85 | 1.0 | 1.0 |
| Rastrigin | 2 / 4| 6 / 12 | 100_000 / 200_000 | 150 / 400 | 1.000 / 0.998 | 1.000 / 0.980 |
| Shubert | 2 / 3 | 18 / 81 | 150_000 / 400_000 | 500 / 1600 | 0.959 / 0.637 | 0.80 / 0.000 |
| Vincent | 2 / 3 | 36 / 216 | 200_000 / 400_000 | 2000 / 4000 | 0.744 / 0.392 | 0.000 / 0.000 |

### End of file